modules/uhttp/client/client.go

```diff
@@ -21,115 +21,53 @@
-		filters = append(filters, tracingFilter(), authenticationFilter(c.info))
-		c.defaultFiltersAdded = true
+func newExecutionChain(
+	filters []Filter, finalTransport http.RoundTripper,
+) executionChain {
+	return executionChain{
+		filters:        filters,
+		finalTransport: finalTransport,
 	}
-	execChain := newExecutionChain(filters, BasicClientFunc(c.do))
-	return execChain.Do(ctx, req)
 }
 
-func (c *Client) do(ctx fx.Context, req *http.Request) (resp *http.Response, err error) {
-	return ctxhttp.Do(ctx, c.Client, req)
+type executionChain struct {
+	currentFilter  int
+	filters        []Filter
+	finalTransport http.RoundTripper
 }
 
-// Get is a context-aware, filter-enabled extension of Get() in http.Client
-func (c *Client) Get(ctx fx.Context, url string) (resp *http.Response, err error) {
-	req, err := http.NewRequest("GET", url, nil)
-	if err != nil {
-		return nil, err
-	}
-
-	return c.Do(ctx, req)
-}
+func (ec executionChain) Send(ctx fx.Context, r *http.Request) (resp *http.Response, err error) {
+	if ec.currentFilter < len(ec.filters) {
+		filter := ec.filters[ec.currentFilter]
+		ec.currentFilter++
 
-// Post is a context-aware, filter-enabled extension of Post() in http.Client
-func (c *Client) Post(
-	ctx fx.Context,
-	url string,
-	bodyType string,
-	body io.Reader,
-) (resp *http.Response, err error) {
-	req, err := http.NewRequest("POST", url, body)
-	if err != nil {
-		return nil, err
+		return filter.Apply(ctx, r, ec)
 	}
-	req.Header.Set("Content-Type", bodyType)
-	return c.Do(ctx, req)
-}
-
-// PostForm is a context-aware, filter-enabled extension of PostForm() in http.Client
-func (c *Client) PostForm(
-	ctx fx.Context,
-	url string,
-	data url.Values,
-) (resp *http.Response, err error) {
-	return c.Post(ctx, url, "application/x-www-form-urlencoded",
-		strings.NewReader(data.Encode()))
-}
 
-// Head is a context-aware, filter-enabled extension of Head() in http.Client
-func (c *Client) Head(ctx fx.Context, url string) (resp *http.Response, err error) {
-	req, err := http.NewRequest("HEAD", url, nil)
-	if err != nil {
-		return nil, err
-	}
-	return c.Do(ctx, req)
+	return ec.finalTransport.RoundTrip(r.WithContext(ctx))
 }
 
-// BasicClient is the simplest, context-aware HTTP client with a single method Do.
-type BasicClient interface {
-	// Do sends an HTTP request and returns an HTTP response, following
-	// policy (e.g. redirects, cookies, auth) as configured on the client.
-	Do(ctx fx.Context, req *http.Request) (resp *http.Response, err error)
-}
-
-// The BasicClientFunc type is an adapter to allow the use of ordinary functions as BasicClient.
-type BasicClientFunc func(ctx fx.Context, req *http.Request) (resp *http.Response, err error)
-
-// Do implements Do from the BasicClient interface
-func (f BasicClientFunc) Do(
-	ctx fx.Context, req *http.Request,
-) (resp *http.Response, err error) {
-	return f(ctx, req)
+func (ec executionChain) RoundTrip(r *http.Request) (resp *http.Response, err error) {
```

Comment:
Why do we need a RoundTrip method on executionChain? Better to stick to the naming from Sender interface. Only execChain finalTransport will invoke the RoundTrip function. Everywhere else, we can have Send or Do